base/task.jl

```diff
@@ -365,6 +365,77 @@ function wait(t::Task)
     nothing
 end
 
+# Wait multiple tasks
+waitany(tasks) = _wait_multiple(tasks)
+waitall(tasks; failfast=false) = _wait_multiple(tasks; all=true, failfast=failfast)
+
+function _wait_multiple(waiting_tasks; all=false, failfast=false)
+    tasks = Task[]
+
+    for (i, t) in enumerate(waiting_tasks)
+        t isa Task || error("Expected an iterator of `Task` object")
+        push!(tasks, t)
+    end
+
+    exception = false
+    nremaining::Int = length(tasks)
+    done_mask = falses(nremaining)
+    for (i, t) in enumerate(tasks)
+        if istaskdone(t)
+            done_mask[i] = true
+            exception |= istaskfailed(t)
+            nremaining -= 1
+        else
+            done_mask[i] = false
+        end
+    end
+
+    if nremaining == 0
+        return tasks, Task[]
+    elseif any(done_mask) && (!all || (failfast && exception))
+        return tasks[done_mask], tasks[.~done_mask]
+    end
+
+    chan = Channel{Tuple{Int,Task}}(Inf)
+    waiter_tasks = Dict{Task, Task}()
+
+    for (i, done) in enumerate(done_mask)
+        if !done
+            t = tasks[i]
+            waiter = @task put!(chan, (i, t))
+            waiter.sticky = false
+            _wait2(t, waiter)
+            waiter_tasks[t] = waiter
+        end
+    end
+
+    while true
+        i, t = take!(chan)
+        delete!(waiter_tasks, t)
+        done_mask[i] = true
+        exception |= istaskfailed(t)
+        nremaining -= 1
+
+        if nremaining == 0 || (!all || failfast && exception)
```

Comment:
another small possible improvement here is to try to empty the channel before breaking, if it can be done non-blocking:
```suggestion
        # stop early if requested, unless there is something already immediately already ready to consume from the channel (using a race-y check)
        if (!all || (failfast && exception)) && !isready(chan)
```